src/mender-update/daemon/context.cpp

```diff
@@ -151,7 +151,11 @@ Context::Context(
 	mender::update::context::MenderContext &mender_context, events::EventLoop &event_loop) :
 	mender_context(mender_context),
 	event_loop(event_loop),
+#ifdef MENDER_USE_DBUS
 	authenticator(event_loop),
+#elif defined(MENDER_EMBED_MENDER_AUTH)
+	authenticator(event_loop, mender_context.GetConfig()),
+#endif
```

Comment:
These kind of ifdefs for constructors I am expecting to see more and more as we do the ports to different platforms. I don't really love it but I understand that implementing a pattern like Abstract Factories (or Factory Method, I always confuse them) is not justified.

Do you have any thoughts on this moving forward? How many variants you think is acceptable for an ifdef until the code gets unmaintainable in your opinion?

Reply:
I don't have a definitive answer to that. A possible solution to the case here is to require the same arguments in both cases, even when it's not needed.

But more generally I would agree that factory functions should be preferred. Hopefully we can avoid many ifdefs with that.

However, one type of ifdef which is unavoidable with the current design, is ifdef when either declaring the type, or when using `make_unique`/`make_shared`. See these examples:

```c++
#ifdef LINUX
KeyValueDatabaseLmdb database_;
#else
KeyValueDatabaseZephyr database_;
#endif

- or -

#ifdef LINUX
auto database = make_shared<KeyValueDatabaseLmdb>();
#else
auto database = make_shared<KeyValueDatabaseZephyr>();
#endif
```

This is because each platform have their own type.

Back when Ole, Vratislav and I were discussing this, we were talking about the alternative where `KeyValueDatabase` would not have different names, but would instead be defined once in each platform section, so you would have multiple `KeyValueDatabase::KeyValueDatabase` constructors for example, once for each platform. The platform name would not be embedded in the name. This would eliminate many ifdefs in the production code, which might be an advantage. I know this is how Qt does it.

In the end we decided against it in favor of using qualified names. I think there was no strong reason in favor of either approach at the time, and it was more a case of not being able to plan everything in detail that far ahead. But I don't think there is a strong reason to stay with this approach, if the alternative makes the code cleaner.